### PR TITLE
Improve factorization speed

### PR DESCRIPTION
**Pull request by mbkma**
Together with my friend [uscpr](https://github.com/uscpr) I had a little play with mate-calc:

- implements Miller-Rabin primality test
- implements Pollard's rho algorithm to find prime factors
- adds mp_new_from_integer function for convenience

Before / After (with PR)

Peek 2020-10-07 21-32

### CODE REVIEW

File: src/mp.c [hunk 1]
```diff
@@ -61,6 +63,14 @@ mp_new(void)
     return z;
 }
 
+MPNumber
+mp_new_from_integer(uint64_t x)
+{
+    MPNumber z;
+    mpc_init2(z.num, PRECISION);
+    mpc_set_si(z.num, x, MPC_RNDNN);
```
Comment (rbuj) [inline src/mp.c:71]: Wrong datatype (sign and size) of 2nd argument passed to mpc_set_si

See page 9 http://www.multiprecision.org/downloads/mpc-1.2.0.pdf

Comment (mbkma) [reply src/mp.c]: thanks, should be fixed now.

**Review by rbuj — APPROVED** [review]
LGTM.

Test 1: prime numbers less than 100
```
100!
93326215443944152681699238856266700490715968264381621468592963895217599993229915608941463976156518286253697920827223758251185210916864000000000000000000000000
2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×2×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×3×5×5×5×5×5×5×5×5×5×5×5×5×5×5×5×5×5×5×5×5×5×5×5×5×7×7×7×7×7×7×7×7×7×7×7×7×7×7×7×7×11×11×11×11×11×11×11×11×11×13×13×13×13×13×13×13×17×17×17×17×17×19×19×19×19×19×23×23×23×23×29×29×29×31×31×31×37×37×41×41×43×43×47×47×53×59×61×67×71×73×79×83×89×97
```
Test 2: consecutive 5-digit prime numbers
```
48157×48271×48397
112503015115159
```
Test 3:
```
./src/test-mp-equation 
Passed all 448 tests
```

**Comment by mbkma** [comment]
I am confused now with the data types. In gnome-calculator they generally use the `int64_t` and cast it to `long` for `mpc_` / `mpfr_` functions. Isn't that undefined behaviour? I mean `int64_t` can be larger than `long`.

https://gitlab.gnome.org/GNOME/gnome-calculator/-/blob/master/vapi/mpfr.vapi#L60-63

https://gitlab.gnome.org/GNOME/gnome-calculator/-/blob/master/lib/number.vala

**Comment by rbuj** [comment]
long is 4 bytes wide on i386, 8 bytes wide on x86_64. Test:
```
$ cat test.c 
/* gcc -Wall test.c -o test */
#include <stdio.h>
#include <stdint.h>

int main (void)
{
   printf ("char %ld\n", sizeof (char));
   printf ("int %ld\n", sizeof (int));
   printf ("long %ld\n", sizeof (long));
   printf ("int64_t %ld\n", sizeof (int64_t));
   printf ("long long %ld\n", sizeof (long long));
   return 0;
}
$ uname -m
x86_64
$ ./test
char 1
int 4
long 8
int64_t 8
long long 8
```

**Comment by raveit65** [comment]
@mbkma 
I just send an invitation to help maintaining the repo.

**Review by raveit65 — APPROVED** [review]
LGTM,
builds and runs fine.

**Comment by raveit65** [comment]
Ready to merge or do we have open questions?

**Comment by mbkma** [comment]
> @mbkma
> I just send an invitation to help maintaining the repo.

thanks

> Ready to merge or do we have open questions?

Please do not merge yet, I want to figure out the correct usage of data types first :)

**Comment by rbuj** [comment]
> Please do not merge yet, I want to figure out the correct usage of data types first :)

@mbkma you can use -Wconversion flag:

make.sh
```
#!/bin/bash
export LANG=en_US.UFT-8

NPROC=`nproc`
CFLAGS="-march=native -mtune=native -fomit-frame-pointer -malign-double -fstrict-aliasing -ffast-math -Wconversion -Wunused-macros -Wunused-parameter"
CONF="--prefix=/usr --enable-compile-warnings=maximum --enable-cxx-warnings=maximum --disable-debug"
CFLAGS="${CFLAGS}" ./autogen.sh ${CONF} &> conf.log && make -j${NPROC} &> make.log && sudo make install &> /dev/null
```
```
$ sh make.sh
$ grep conversion mate-calc/make.log 
math-buttons.c:566:91: warning: conversion to 'guint' {aka 'unsigned int'} from 'gint' {aka 'int'} may change the sign of the result [-Wsign-conversion]
math-buttons.c:912:62: warning: conversion to 'long unsigned int' from 'gint' {aka 'int'} may change the sign of the result [-Wsign-conversion]
math-buttons.c:916:60: warning: conversion to 'long unsigned int' from 'gint' {aka 'int'} may change the sign of the result [-Wsign-conversion]
math-buttons.c:955:62: warning: conversion to 'long unsigned int' from 'gint' {aka 'int'} may change the sign of the result [-Wsign-conversion]
math-buttons.c:959:60: warning: conversion to 'long unsigned int' from 'gint' {aka 'int'} may change the sign of the result [-Wsign-conversion]
/usr/lib64/glib-2.0/include/glibconfig.h:98:29: warning: conversion to 'guint' {aka 'unsigned int'} from 'int' may change the sign of the result [-Wsign-conversion]
/usr/lib64/glib-2.0/include/glibconfig.h:98:29: warning: conversion to 'guint' {aka 'unsigned int'} from 'int' may change the sign of the result [-Wsign-conversion]
mp.c:425:17: warning: conversion to 'ulong' {aka 'long unsigned int'} from 'long int' may change the sign of the result [-Wsign-conversion]
mp.c:430:13: warning: conversion to 'ulong' {aka 'long unsigned int'} from 'long int' may change the sign of the result [-Wsign-conversion]
mp.c:494:42: warning: conversion to 'long unsigned int' from 'long int' may change the sign of the result [-Wsign-conversion]
mp-binary.c:56:15: warning: conversion from 'size_t' {aka 'long unsigned int'} to 'int' may change value [-Wconversion]
mp-binary.c:57:15: warning: conversion from 'size_t' {aka 'long unsigned int'} to 'int' may change value [-Wconversion]
mp-convert.c:174:18: warning: conversion from 'size_t' {aka 'long unsigned int'} to 'int' may change value [-Wconversion]
mp-convert.c:241:42: warning: conversion from 'size_t' {aka 'long unsigned int'} to 'int' may change value [-Wconversion]
mp-equation.c:154:14: warning: conversion from 'int' to 'char' may change value [-Wconversion]
mp-equation.c:200:14: warning: conversion from 'int' to 'char' may change value [-Wconversion]
mp-equation.c:313:11: warning: conversion to 'int' from 'guint' {aka 'unsigned int'} may change the sign of the result [-Wsign-conversion]
financial.c:70:11: warning: conversion from 'long int' to 'int' may change value [-Wconversion]
prelexer.c:18:21: warning: conversion from 'size_t' {aka 'long unsigned int'} to 'guint' {aka 'unsigned int'} may change value [-Wconversion]
prelexer.c:42:29: warning: conversion from 'long int' to 'guint' {aka 'unsigned int'} may change value [-Wconversion]
/usr/include/glib-2.0/glib/gunicode.h:768:29: warning: conversion from 'long int' to 'guint' {aka 'unsigned int'} may change value [-Wconversion]
parserfunc.c:199:39: warning: conversion to 'size_t' {aka 'long unsigned int'} from 'long int' may change the sign of the result [-Wsign-conversion]
parserfunc.c:260:39: warning: conversion to 'size_t' {aka 'long unsigned int'} from 'long int' may change the sign of the result [-Wsign-conversion]
parserfunc.c:604:15: warning: conversion from 'long int' to 'gint' {aka 'int'} may change value [-Wconversion]
parser.c:350:39: warning: conversion to 'size_t' {aka 'long unsigned int'} from 'long int' may change the sign of the result [-Wsign-conversion]
mate-calc-cmd.c:85:32: warning: conversion from 'size_t' {aka 'long unsigned int'} to 'int' may change value [-Wconversion]
mp-serializer.c:126:21: warning: conversion from 'gsize' {aka 'long unsigned int'} to 'int' may change value [-Wconversion]
mp-serializer.c:138:13: warning: conversion from 'long int' to 'int' may change value [-Wconversion]
mp-serializer.c:143:29: warning: conversion from 'gsize' {aka 'long unsigned int'} to 'int' may change value [-Wconversion]
mp-serializer.c:150:35: warning: conversion to 'gsize' {aka 'long unsigned int'} from 'int' may change the sign of the result [-Wsign-conversion]
math-equation.c:1121:14: warning: conversion from 'int' to 'char' may change value [-Wconversion]
math-equation.c:1144:14: warning: conversion from 'int' to 'char' may change value [-Wconversion]
```

**Comment by mbkma** [comment]
ok, should be good now

**Comment by rbuj** [comment]
There are still some conversion warnings... 
```diff
< -Wconversion 31
> -Wconversion 19
< -Wshadow 2
< -Wsign-conversion 14
> -Wsign-conversion 10
```
```diff
diff --git a/src/financial.c b/src/financial.c
index 337ff93..7c7cbf0 100644
--- a/src/financial.c
+++ b/src/financial.c
@@ -60,8 +60,8 @@ calc_ddb(MathEquation *equation, MPNumber *t, MPNumber *cost, MPNumber *life, MP
  *
  */
 
-    int i;
-    int len;
+    long i;
+    long len;
     MPNumber MPbv = mp_new();
     MPNumber MP1 = mp_new();
     MPNumber MP2 = mp_new();
diff --git a/src/mate-calc-cmd.c b/src/mate-calc-cmd.c
index cf8db3d..d8e86ec 100644
--- a/src/mate-calc-cmd.c
+++ b/src/mate-calc-cmd.c
@@ -70,7 +70,6 @@ int
 main(int argc, char **argv)
 {
     char *equation, *line;
-    size_t nbytes = MAXLINE;
 
     /* Seed random number generator. */
     srand48((long) time((time_t *) 0));
@@ -82,7 +81,7 @@ main(int argc, char **argv)
     equation = (char *) malloc(MAXLINE * sizeof(char));
     while (1) {
         printf("> ");
-        line = fgets(equation, nbytes, stdin);
+        line = fgets(equation, MAXLINE, stdin);
 
         if (line != NULL)
             str_adjust(equation);
diff --git a/src/mp-convert.c b/src/mp-convert.c
index 7626a4a..40fcf1d 100644
--- a/src/mp-convert.c
+++ b/src/mp-convert.c
@@ -116,7 +116,8 @@ mp_to_double(const MPNumber *x)
 static int
 char_val(char **c, int base)
 {
-    int i, j, value, offset;
+    int i, j, value;
+    size_t offset;
     const char *digits[][10] = {{"٠", "١", "٢", "٣", "٤", "٥", "٦", "٧", "٨", "٩"},
                                 {"〇", "〡", "〢", "〣", "〤", "〥", "〦", "〧", "〨", "〩"},
                                 {"۰", "۱", "۲", "۳", "۴", "۵", "۶", "۷", "۸", "۹"},
@@ -196,7 +197,7 @@ ends_with(const char *start, const char *end, const char *word)
 
 // FIXME: Doesn't handle errors well (e.g. trailing space)
 static bool
-set_from_sexagesimal(const char *str, int length, MPNumber *z)
+set_from_sexagesimal(const char *str, size_t length, MPNumber *z)
 {
     int degrees = 0, minutes = 0;
     char seconds[length+1];
diff --git a/src/mp-serializer.c b/src/mp-serializer.c
index 3967cbc..e353ea3 100644
--- a/src/mp-serializer.c
+++ b/src/mp-serializer.c
@@ -63,7 +63,8 @@ mp_to_string_real(MpSerializer *serializer, const MPNumber *x, int base, gboolea
     MPNumber integer_component = mp_new();
     MPNumber fractional_component = mp_new();
     MPNumber temp = mp_new();
-    int i, last_non_zero;
+    int i;
+    gsize last_non_zero;
 
     if (mp_is_negative(x))
         mp_abs(x, &number);
@@ -92,7 +93,7 @@ mp_to_string_real(MpSerializer *serializer, const MPNumber *x, int base, gboolea
         MPNumber t = mp_new();
         MPNumber t2 = mp_new();
         MPNumber t3 = mp_new();
-        int64_t d;
+        long d;
 
         if (serializer->priv->base == 10 && serializer->priv->show_tsep && i == serializer->priv->tsep_count) {
             g_string_prepend_unichar(string, serializer->priv->tsep);
@@ -130,7 +131,7 @@ mp_to_string_real(MpSerializer *serializer, const MPNumber *x, int base, gboolea
     /* Write out the fractional component */
     mp_set_from_mp(&fractional_component, &temp);
     for (i = serializer->priv->trailing_digits; i > 0 && !mp_is_zero(&temp); i--) {
-        int d;
+        long d;
         MPNumber digit = mp_new();
 
         mp_multiply_integer(&temp, base, &temp);
diff --git a/src/mp.c b/src/mp.c
index 27ea8a8..dee945c 100644
--- a/src/mp.c
+++ b/src/mp.c
@@ -408,23 +408,23 @@ mp_reciprocal(const MPNumber *x, MPNumber *z)
 }
 
 void
-mp_root(const MPNumber *x, int64_t n, MPNumber *z)
+mp_root(const MPNumber *x, long n, MPNumber *z)
 {
-    uint64_t p;
+    ulong p;
 
     if (n < 0)
     {
         mpc_ui_div(z->num, 1, x->num, MPC_RNDNN);
 
-        if (n == INT64_MIN)
-            p = (uint64_t) INT64_MAX + 1;
+        if (n == LONG_MIN)
+            p = (ulong) LONG_MAX + 1;
         else
-            p = -n;
+            p = (ulong) -n;
     }
     else if (n > 0)
     {
         mp_set_from_mp(x, z);
-        p = n;
+        p = (ulong) n;
     }
     else
     {   /* Translators: Error displayed when attempting to take zeroth root */
@@ -434,14 +434,14 @@ mp_root(const MPNumber *x, int64_t n, MPNumber *z)
     }
     if (!mp_is_complex(x) && (!mp_is_negative(x) || (p & 1) == 1))
     {
-        mpfr_rootn_ui(mpc_realref(z->num), mpc_realref(z->num), (ulong) p, MPFR_RNDN);
+        mpfr_rootn_ui(mpc_realref(z->num), mpc_realref(z->num), p, MPFR_RNDN);
         mpfr_set_zero(mpc_imagref(z->num), MPFR_RNDN);
     }
     else
     {
         mpfr_t tmp;
         mpfr_init2(tmp, PRECISION);
-        mpfr_set_ui(tmp, (ulong) p, MPFR_RNDN);
+        mpfr_set_ui(tmp, p, MPFR_RNDN);
         mpfr_ui_div(tmp, 1, tmp, MPFR_RNDN);
         mpc_pow_fr(z->num, z->num, tmp, MPC_RNDNN);
         mpfr_clear(tmp);
@@ -687,7 +687,7 @@ mp_is_pprime(MPNumber *n, ulong rounds)
         }
 
         bool is_witness = FALSE;
-        for (int i = 1; i < l; i++)
+        for (long it = 1; it < l; it++)
         {
             mp_modular_exponentiation(&b, &two, n, &b);
             if (mp_compare(&b, &t) == 0)
diff --git a/src/mp.h b/src/mp.h
index ac89249..22bc986 100644
--- a/src/mp.h
+++ b/src/mp.h
@@ -195,7 +195,7 @@ void   mp_get_eulers(MPNumber *z);
 void   mp_get_i(MPNumber *z);
 
 /* Sets z = n√x */
-void   mp_root(const MPNumber *x, int64_t n, MPNumber *z);
+void   mp_root(const MPNumber *x, long n, MPNumber *z);
 
 /* Sets z = √x */
 void   mp_sqrt(const MPNumber *x, MPNumber *z);
@@ -213,7 +213,7 @@ void mp_modular_exponentiation(const MPNumber *x, const MPNumber *y, const MPNum
 void   mp_xpowy(const MPNumber *x, const MPNumber *y, MPNumber *z);
 
 /* Sets z = x^y */
-void   mp_xpowy_integer(const MPNumber *x, int64_t y, MPNumber *z);
+void   mp_xpowy_integer(const MPNumber *x, long y, MPNumber *z);
 
 /* Sets z = e^x */
 void   mp_epowy(const MPNumber *x, MPNumber *z);
@@ -262,10 +262,10 @@ float  mp_to_float(const MPNumber *x);
 double mp_to_double(const MPNumber *x);
 
 /* Returns x as a native integer */
-int64_t mp_to_integer(const MPNumber *x);
+long mp_to_integer(const MPNumber *x);
 
 /* Returns x as a native unsigned integer */
-uint64_t mp_to_unsigned_integer(const MPNumber *x);
+ulong mp_to_unsigned_integer(const MPNumber *x);
 
 /* Sets z = sin x */
 void   mp_sin(const MPNumber *x, MPAngleUnit unit, MPNumber *z);
diff --git a/src/parser.c b/src/parser.c
index b8cc98f..9d37dc6 100644
--- a/src/parser.c
+++ b/src/parser.c
@@ -347,7 +347,7 @@ p_check_variable(ParserState* state, gchar* name)
         for(c = name; *c != '\0'; c = next)
         {
             next = utf8_next_char(c);
-            snprintf(buffer, next - c + 1, "%s", c);
+            snprintf(buffer, (size_t)(next - c + 1), "%s", c);
             if(!(*(state->get_variable))(state, buffer, &temp))
             {
                 result = 0;
diff --git a/src/parserfunc.c b/src/parserfunc.c
index c63ff96..6f27daf 100644
--- a/src/parserfunc.c
+++ b/src/parserfunc.c
@@ -584,7 +584,7 @@ void*
 pf_do_x_pow_y_int(ParseNode* self)
 {
     MPNumber* val;
-    gint pow;
+    long pow;
     MPNumber* ans = mp_new_ptr();
     val = (MPNumber*) (*(self->left->evaluate))(self->left);
 
```

**Comment by mbkma** [comment]
@rbuj done

**Comment by rbuj** [comment]
@mbkma thank you so much.

**Pull request merged**